xmtp_mls/src/groups/sync.rs

```diff
@@ -283,41 +293,46 @@ impl MlsGroup {
                 if !allow_epoch_increment {
                     return Err(MessageProcessingError::EpochIncrementNotAllowed);
                 }
-                let maybe_pending_commit = openmls_group.pending_commit();
-                // We don't get errors with merge_pending_commit when there are no commits to merge
-                if maybe_pending_commit.is_none() {
-                    debug!(
-                        "no pending commit to merge. Group epoch: {}. Message epoch: {}",
-                        group_epoch, message_epoch
-                    );
-                    conn.set_group_intent_to_publish(intent.id)?;
-
-                    // Return OK here, because an error will roll back the transaction
-                    return Ok(());
+                if let Some(published_in_epoch) = intent.published_in_epoch {
+                    let published_in_epoch_u64 = published_in_epoch as u64;
+                    let group_epoch_u64 = group_epoch.as_u64();
+
+                    if published_in_epoch_u64 != group_epoch_u64 {
+                        conn.set_group_intent_to_publish(intent.id)?;
+                        log::warn!(
+                            "Intent was published in epoch {} but group is currently in epoch {}",
+                            published_in_epoch_u64,
+                            group_epoch_u64
+                        );
+                        return Ok(());
+                    }
                 }
+
+                let pending_commit = if let Some(staged_commit) = intent.staged_commit {
+                    decode_staged_commit(staged_commit)?
+                } else {
+                    return Err(MessageProcessingError::IntentMissingStagedCommit);
+                };
+
                 log::info!(
                     "[{}] Validating commit for intent {}. Message timestamp: {}",
                     self.context.inbox_id(),
                     intent.id,
                     envelope_timestamp_ns
                 );
+
                 let maybe_validated_commit = ValidatedCommit::from_staged_commit(
                     client,
                     conn,
-                    maybe_pending_commit.expect("already checked"),
+                    &pending_commit,
                     openmls_group,
                 )
                 .await;
 
                 if maybe_validated_commit.is_err() {
-                    log::warn!("error validating commit: {:?}", maybe_validated_commit);
```

Comment:
Should we keep this log line? Think it might be important. It's unclear if this scenario is expected or not, my assumptions:
1. We validate the commit at the time we created the intent
2. If we are here, it means no new commits have happened since we last validated it

Therefore it seems unexpected, and should be logged as an error?

Separately to this PR, another case I'm unsure if we handle or not - we want to add [A, B] to the group, but somebody else adds [B, C] and their commit comes in before ours. Ideally we still add [A], even though [B] is no longer valid to be added.

Reply:
I don't think we have a test for that, but I think it _should_ work. Worth investigating to find out.

I'll add the log line back in. Got a little aggressive pruning old code.